Add additional warnings for lint rules

daghack opened the pull request:
We should define a set of rules for evaluating a good or bad Dockerfile. Create a method for evaluating the Dockerfile to ensure obvious issues are highlighted to the user.

This PR adds 3 additional lint warnings for the Dockerfile frontend:
- lowercased stage names
- lowercased command flag values
- consistently cased commands

These rules are definitely not the comprehensive list of linting warnings we should add, but instead represent an initial set, with more to come in additional PRs moving forward! 😄 

( This PR existed in a similar, previous PR #4751, but only implements the warnings. The additional `--print=lint` functionality will come in its own PR. 😄 )

thaJeztah commented on frontend/dockerfile/dockerfile2llb/convert_test.go:
```diff
@@ -27,6 +27,57 @@ func toEnvMap(args []instructions.KeyValuePairOptional, env []string) map[string
+	assert.Empty(t, warningList)
+
+	df = `FROM scratch AS FOO
+ENV FOO bar
```
Oh! And if we consider "best practices", we could consider that a Dockerfile should start with a `# syntax`☺️

daghack replied:
Those are great suggestions! There's quite a few linting rules that we want to add moving forward, these three are meant to be just the initial set, with other linting rules being introduced in future PRs! I've updated the initial comment to note that, because this is definitely just a first stepping stone! :D

thaJeztah replied:
For sure! I saw your PR and was interested; consider my comment purely as "hey, these might be interesting to add as well!"

colinhemmings replied:
These are great, thanks @thaJeztah. I will send out a proposal shortly to capture more input on this.

daghack commented:
I wasn't content with how it was printing, so I refactored and added an additional callback for fetching source information. And because I forgot to include this prior, attached are screenshots for the end result. 😄 

![Screenshot 2024-03-14 at 4 07 06 PM](https://github.com/moby/buildkit/assets/1680823/fa962eb3-dc3a-49e7-be30-d580e1c3e5f1)
![Screenshot 2024-03-14 at 4 07 38 PM](https://github.com/moby/buildkit/assets/1680823/dcb0a4b6-63de-4845-9f08-078898a19fc2)

daghack commented:
![Screenshot 2024-03-14 at 4 32 11 PM](https://github.com/moby/buildkit/assets/1680823/d3b050ca-1c79-4b7f-b989-e20572c0bf60)
![Screenshot 2024-03-14 at 4 32 44 PM](https://github.com/moby/buildkit/assets/1680823/c16c6246-775f-4794-8e62-a5a81481e00b)
![Screenshot 2024-03-14 at 4 36 10 PM](https://github.com/moby/buildkit/assets/1680823/254b713d-88ff-475f-83b4-809a5192f639)

daghack commented:
I'm not a super fan of the lack of any real context in the inline, and think that perhaps having the output in those cases include filename+linenumbers, but when I tried this it came at the cost of the `--debug` info looking pretty cluttered and having redundant information between the Short and the additional printed information.

@tonistiigi Thoughts/Preferences

tonistiigi commented:
@daghack We can improve the short message so that instead of constant text it says "prefix: FOObar casing is inconsistent with the rest of the Dockerfile (app.Dockerfile:12)" or smth similar.

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -109,12 +110,51 @@ func Dockefile2Outline(ctx context.Context, dt []byte, opt ConvertOpt) (*outline
+	return
+}
+
+func DockerfileLint(ctx context.Context, dt []byte, opt ConvertOpt) ([]linter.Warning, error) {
```
This function belongs more to #4761 I think

daghack replied:
Agreed! Kept it in for testing purposes, but with the testing related comment below, going to refactor those, so this can get the boot from here!

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert_test.go:
```diff
@@ -27,6 +27,57 @@ func toEnvMap(args []instructions.KeyValuePairOptional, env []string) map[string
 	return m
 }
 
+func TestDockerfileLinting(t *testing.T) {
```
We can either add both unit and integration tests for the rules or only integration tests. I see in the other PR you started to add integration tests as well.

Ideally we make it so that a test case has a Dockerfile, and expected warnings. Then integration test variant would build it and check for expected warnings in progress (we should have existing test for that can be modified) and another variant does the same via `--print=lint` and checks the expected warnings in output.

tonistiigi commented on frontend/dockerfile/linter/ruleset.go:
```diff
@@ -0,0 +1,110 @@
+	"github.com/moby/buildkit/frontend/dockerfile/parser"
+)
+
+const (
```
Make sure we add the existing line-continuation parser warning also to this list.

tonistiigi commented on frontend/dockerfile/instructions/parse.go:
```diff
@@ -84,6 +92,11 @@ func ParseInstruction(node *parser.Node) (v interface{}, err error) {
 		return parseCopy(req)
 	case command.From:
+		if !linter.ValidateStageNameCasing(req.args) {
+			location := linter.FirstNodeRange(node)
```
Not sure why there is the extra function to get the first range instead of just returning all.

daghack replied:
`opt.Warn` doesn't take a list of ranges, it takes a single range. I figured grabbing the first range was sensible in that case.

tonistiigi replied:
Looks like the array goes missing in conversion in https://github.com/moby/buildkit/blob/master/frontend/dockerfile/builder/build.go#L244-L253 . The actual type in API is array.

daghack replied:
Cool, will fix that. 👍

tonistiigi commented on frontend/dockerfile/instructions/parse.go:
```diff
@@ -66,8 +69,13 @@ func newParseRequestFromNode(node *parser.Node) parseRequest {
+}
+
+// ParseInstruction converts an AST to a typed instruction (either a command or a build stage beginning when encountering a `FROM` statement)
+func ParseInstructionWithWarnings(node *parser.Node, warn warnCallback) (v interface{}, err error) {
```
Private types in public methods should be avoided.

It may make sense to turn these to methods of `Parser` that can have the warner configured so it doesn't need to be passed all the time.

daghack replied:
I ended up refactoring this as part of getting the better short messages in, but this is a really good callout for me, as I hadn't even considered that.

tonistiigi commented on frontend/dockerfile/instructions/parse.go:
```diff
@@ -84,6 +92,11 @@ func ParseInstruction(node *parser.Node) (v interface{}, err error) {
+		if !linter.ValidateStageNameCasing(req.args) {
+			location := linter.FirstNodeRange(node)
+			stageNameCasing := linter.LinterRules[linter.RuleStageNameCasing]
+			warn(stageNameCasing.Short, "", [][]byte{[]byte(stageNameCasing.Description)}, location)
```
I'd like this to be so the function that triggers the warning only accepts the type from the `linter` pkg and the values defined there are included automatically instead of reading from a map and then passing individual values. In here, only location and maybe something about the textual value of "node" is added.

tonistiigi commented on frontend/dockerfile/linter/ruleset.go:
```diff
@@ -0,0 +1,110 @@
+
+const (
+	RuleStageNameCasing = "StageNameCasing"
+	RuleFlagCasing      = "FlagCasing"
```
`RuleFlagCasing` was a mistake I think. Looks like the flags are already case sensitive.

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -109,12 +110,51 @@ func Dockefile2Outline(ctx context.Context, dt []byte, opt ConvertOpt) (*outline
+			Short:    short,
+			Detail:   detail,
+			Location: location,
+			Filename: filename,
```
We should try to keep the `SourceInfo` type in here. Maybe even list of warnings ans list of sourceinfo and then warning has a sourceinfo index. This is what we do in provenance for example for each build step.

tonistiigi commented on frontend/dockerfile/linter/ruleset.go:
```diff
@@ -0,0 +1,110 @@
+	return consistentCasing(cmd)
+}
+
+func ValidateNodeLintRules(node *parser.Node) []parser.Warning {
```
I think this actual validation logic belongs to the caller side. Atm. we have 2 rules in linter package and 1 in instructions pkg. Better to keep the linter pkg only for the definitions and helpers for triggering warnings rather than include actual build logic over `parser.Node` etc. in here.

daghack commented:
Ended up being a bit of refactoring to get cleaner output, and updated to address @tonistiigi feedback!
Screenshots of updated output (the first being what displays inline during the build, the with the additional context displayed via the `docker --debug` flag.

![Screenshot 2024-03-18 at 8 42 30 AM](https://github.com/moby/buildkit/assets/1680823/b25c7c42-0806-4fb9-9493-8d2269a4b0f0)
![Screenshot 2024-03-18 at 8 43 24 AM](https://github.com/moby/buildkit/assets/1680823/b9ed380e-19b4-4113-acf8-ed29aa979676)

daghack commented:
Incorporated lots of feedback by @tonistiigi, now looking for a review! :)

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -180,13 +186,31 @@ func toDispatchState(ctx context.Context, dt []byte, opt ConvertOpt) (*dispatchS
 
-	for _, w := range dockerfile.Warnings {
-		opt.Warn(w.Short, w.URL, w.Detail, w.Location)
+	// Moby still uses the `dockerfile.PrintWarnings` method to print warnings,
```
follow-up: we can change the Moby codebase as well if needed. We can't just change the implementation so the printing that Moby does becomes impossible to implement.

tonistiigi replied:
follow-up: we can change the Moby codebase as well if needed. We can't just change the implementation so the printing that Moby does becomes impossible to implement.

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -180,13 +186,31 @@ func toDispatchState(ctx context.Context, dt []byte, opt ConvertOpt) (*dispatchS
+		if warning.URL == linter.RuleNoEmptyContinuations.URL {
+			location := []parser.Range{*warning.Location}
+			msg := linter.RuleNoEmptyContinuations.Format()
+			linter.RuleNoEmptyContinuations.Run(opt.Warn, location, msg)
```
nit. Maybe it would be more readable if organized to:
```
opt.Warn(linter.RuleNoEmptyContinuations, location, msg)
```
or 
```
linter := ... opt.Warn
linter.Show(lintrules.NoEmptyContinuations, location, msg)
```

daghack replied:
My pitch is that we let this be as is for this PR, and then let this + updating `moby`/`PrintWarnings` be a follow up PR.

daghack replied:
@tonistiigi Organizing the code this way basically does not square with how we're currently tying the rule message templates to the rule through generics—Go doesn't support parametrized methods, so neither

`type ConvertOpts struct {
....
    Warn: func[F any](LinterRule[F], ...)
}`

nor
`type Linter struct {
....
}
func (l *Linter)Show[F any](LinterRule[F}, ...) { ... }
`
are going to work because, in either case, the fix is to parameterize the containing struct with our `[F any]`, which is absolutely horribly terribly awfully nonfunctional in the case of `ConvertOpts` and merely ugly and ungainly in the case of a new `Linter` struct.

tonistiigi commented on frontend/dockerfile/dockerfile_test.go:
```diff
@@ -176,6 +176,7 @@ var allTests = integration.TestFuncs(
 	testDockerIgnoreMissingProvenance,
 	testSBOMScannerArgs,
 	testMultiPlatformWarnings,
+	testLintWarnings,
```
I think we should have test for each rule and internally cases that cause warnings and cases that don't. Otherwise this will grow very large.

The shared part could still be the same and not duplicated.

Eg.

```
TestLintConsistentCasing() {
cases := []lintTestCase{{ dockerfile: , expectedWarnings: []})
testWarnings(cases)
// testPrintLint(cases)
}
```

tonistiigi commented on frontend/dockerfile/dockerfile_test.go:
```diff
@@ -6764,6 +6765,102 @@ ARG BUILDKIT_SBOM_SCAN_STAGE=true
+
+# set of commands which should not cause warnings
+FROM scratch as base
+copy Dockerfile .
```
I think this is still warning:

No warning:
```
FROM scratch AS base
COPY
```

```
from scratch as base
copy
```

With warning:

```
FROM scratch as base
COPY
```

```
FROM scratch AS base
copy
```

daghack replied:
Ended up being easiest to make this its own lint rule!

daghack commented:
@tonistiigi Ping :)

tonistiigi commented on frontend/dockerfile/instructions/parse.go:
```diff
@@ -84,6 +89,14 @@ func ParseInstruction(node *parser.Node) (v interface{}, err error) {
 	case command.Copy:
 		return parseCopy(req)
 	case command.From:
+		if lintWarn != nil && !isValidStageNameCasing(req.args) {
```
nit: "isValidStageNameCasing" can just be "isLowerCaseStageName"

tonistiigi commented on frontend/dockerfile/instructions/parse.go:
```diff
@@ -815,3 +828,40 @@ func allInstructionNames() []string {
 	return out
 }
+
+func isLowerCased(s string) bool {
```
nit: I don't think `isLowerCased` and `isUpperCased` are needed. This is simple enough to be inline and more readable without extra wrapping.

tonistiigi replied:
```
type casing int

const (
	caseAny casing = iota
	caseLower
	caseUpper
)

func checkCase(s string, exp casing) (consistent, isLower bool) {
	expLower := exp == caseLower
	for _, r := range s {
		if l := unicode.IsLower(r); exp == caseAny {
			expLower = l
			exp = caseLower
		} else if expLower != l {
			return false, false
		}
	}
	return true, expLower
}
```

could maybe also be an option but probably unnecessary

tonistiigi commented on frontend/dockerfile/dockerfile_test.go:
```diff
@@ -6764,78 +6767,138 @@ ARG BUILDKIT_SBOM_SCAN_STAGE=true
-		}
-	}()
+# warning: 'copy' should match first command's casing (in this case, 'FROM's casing)
+copy Dockerfile /foo
```
Add another lowercase command as well to test that this warning only appears once per file and if I have `FROM` followed by 20 `copy/run` then I don't get 20 warnings.

daghack replied:
I don't think that would be correct behavior. It's maybe noisy if there's a `FROM` followed by 20 `copy/run`, but having it happen only once per file means that if I have, say, two instances where I have erroneously used alternate casing, then I could lint it, fix EVERY lint warning it reports, and then STILL fail the subsequent lint, but now with different warnings.

If I had, say, 3 inconsistent commands out of 100, it would be maddening to have to run the linter 3 different times in order to be able to have all of these reported.

I'm going to opt instead to be more sophisticated in how we check for consistent command casing throughout the file to compare whether a command matches the most used casing, which will prevent this scenario!

tonistiigi commented on frontend/dockerfile/dockerfile_test.go:
```diff
@@ -6962,6 +7025,82 @@ RUN rm -f /foo-2030.1
+		t.Fatalf("timed out waiting for statusDone")
+	}
+
+	<-statusDone
```
Not needed

daghack replied:
Cool! That was just copy / pasted from the prior example of how to use the integration function for these tests, so I can clean it up!

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -180,13 +182,46 @@ func toDispatchState(ctx context.Context, dt []byte, opt ConvertOpt) (*dispatchS
+		}
+	}
+
+	var firstConsistentCommand string
```
Maybe move this block with the `range dockerfile.AST.Children` to a helper function `validateConsistentCasing(dockerfile, opt.Warn)`

daghack replied:
Absolutely a great idea, the check logic definitely built up here.

daghack commented:
Getting there, slowly but surely!

daghack commented:
@tonistiigi 🎉 No write access, can I get a merge too? 😮 Ty!
